### ISSUE TITLE
safe map maybe mapOr?

### ISSUE THREAD

**Issue by jon49** [ramda/ramda#1389]
Could it be possible to create a safe map that doesn't blow up if it's not an array. Similar to `propOr`.

E.g.,

``` js
const mapOr = r.curry((default, f, a) =>
    r.is(Array, a)
        ? r.map(f, a)
    : r.map(f, default)
)
```

Let me know if you would like me to write a proper `mapOr` function.

**Comment by davidchambers** [comment]
One should know the shape of one's data. If one's dealing with a source which provides data in one of two shapes, one should normalize immediately rather than rely on one's functions to handle both forms.

**Comment by jon49** [comment]
Couldn't one say the same for `propOr`?

In all seriousness, I think in an ideal world that would be true. But since we are using a loosely typed language -- JavaScript -- it becomes quite difficult. I use TypeScript to help, but it only helps superficially. In other languages they might have things like `options` or some static analysis before compiling the code.

I think the real use case comes when we are writing abstractions on data and it gets clunky testing all the time.

I think my ideal would be that `map` itself would be safe (returning an empty array) but I can see many cases where you would like the code to blow up.

So, in summary, although I agree in strongly typed languages we should know what the data will be exactly. In JavaScript it isn't the case.

**Comment by davidchambers** [comment]
My view is that even in a dynamically typed language such as JavaScript our programs should be well typed (i.e. functions should always be applied to the correct number of arguments of the correct types). It takes more discipline to achieve this in a dynamically typed language than in a statically typed language, but this shouldn't prevent us from trying.

My concern is that having functions "work" when given input of the wrong type makes it *more* difficult to write well-typed programs. I would rather have the program explode so it's clear that something is wrong.

I'm just one of many Ramda contributors, and I'm at the "Haskell" end of the spectrum. Others may support this proposal.

**Comment by jon49** [comment]
Another example of `Ramda` being loose is the function `path`. The path doesn't necessarily need to be there and it just returns `undefined`. I think this also shows that in JavaScript at least, uncertainty is something that is necessarily embraces and that is what makes it such a dynamic language. In other languages certainty may be of higher value, but I don't know if that is necessarily reflective of JavaScript.

**Comment by CrossEye** [comment]
Funny, I thought I responded to this before.

I see a big difference between the suggestion and `propOr` or `path`.  Those ones handle failures in processing the _contents_ of the data.  The suggestion here is to deal with the _type_ of what's supplied.  That's where I've generally drawn the line.

So, for instance, why do you test that the third parameter is an array, but not that the second is a function?  Why not check that the function handles the types of elements included in the array?  Should we check that the elements all share a common type?  And how about the default?  In case the second element is not a function, should the default have the same shape as the array parameter?  What if the default is `undefined`?  Is that ok to return?

These sorts of questions are why I tend to prefer to assume that users should be able to make these decisions themselves, and that Ramda doesn't really want to get into the type-checking business.  I might one day be convinced otherwise.  (David Chambers is working on that.)  But adding a function like this, whose whole point is just to type-check something and offer an alternate process if the type isn't the normal one seems pretty far from Ramda's core.

**Comment by jon49** [comment]
I would be OK with a `safeMap` that just check to make sure that there is something there, e.g.,

```ts
R.map(R.identity, void 0) // => []
```

This would remove the "type checking" problem and make it more of a processing of the *contents*. E.g.,

```ts
R.pipe(
  R.path(['some', 'nice', 'path']),
  R.map(R.identity)
)(xs)
// where xs = {some: {not: {soNice: 'path'}}}
// => []
```

I don't see the difference in erring out at `path` and `map` but it seems odd to error out at just `map`.

**Comment by CrossEye** [comment]
To me, there's not really much difference between checking whether the parameter is `undefined`/`null` or not and checking whether it has the correct type.  I believe @davidchambers would suggest that the correct answer would be to throw an error such as `TypeError: ‘map’ requires a list as its second parameter.  Received 'undefined'.`.

I suppose I can see getting to that point one day, but I'm not looking forward to the code that does so.

**Comment by CrossEye** [comment]
> I don't see the difference in erring out at `path` and `map` but it seems odd to error out at just `map`.

I'm not sure I understand what you mean.  We don't throw any errors (at least not intentionally) in `path`.  We do offer a fairly graceful exit, in the case that the full path is not defined, in returning `undefined`, but that is the most logical answer to the question in any case.  `map` does not have a graceful fallback, so we don't try to synthesize one.

I understand that this request is an attempt to solve a real problem.  The fact that `map` can throw makes it much more difficult to compose properly.  But I think if we're going to solve that problem, we need to solve it more universally than just addressing `map`.

**Comment by jon49** [comment]
What I meant by erring out both `path` and `map` was that throwing an error on both of those would be logical. But it doesn't seem logical to throw an error on one but not the other. It seems inconsistent.

As in the last example I showed. One might expect to find an array in a nested object but get undefined. Rather than throwing on map it seems to make sense just to return an empty array. This would make the code "self healing." In that way. Rather similar to how browsers are rather lenient. I think that is part of the beauty of JavaScript, in that it is more forgiving. It we want a language that is more exact then we could use something like PureScript.

I'm not sure how to make checking for undefined more universal.

Any ways, that's my two cents. :-)

**Comment by CrossEye** [comment]
The contract for `map`, in my opinion, is to apply a function to a container, and return a new container of the same shape with that function applied to its contents.  So

```js
map(square, [1, 2, 3]); //=> [1, 4, 9]
map(square, Just(5)); //=> Just(25)
map(square, Nothing); //=> Nothing
map(square, Tree(Tree(Node(1), Node(2)), Tree(Tree(Node(3), Tree(Node(4), Node(5))))));
//=> Tree(Tree(Node(1), Node(4)), Tree(Tree(Node(9), Tree(Node(16), Node(25)))))
```

Most of this is handled through delegation.  But if your type has a `map` function on it, we'll call it.

So I think there are reasonable questions about what we should do if `map` is passed something that has no `map` method and is not a list.  What if that parameter is an object?  A boolean? A string? (That one actually "works".)  How about `null` / `undefined`?

I would love to come to some sort of reasonable conclusion about this.  And I really would like it if the conclusion didn't involve throwing an error.  But I don't see any reasonable alternative.

So I'd love it if you keep throwing in your two+ cents.  This is something we really need to address.

**Comment by raine** [comment]
> This would remove the "type checking" problem and make it more of a processing of the contents. E.g.,
> ```js
R.pipe(
  R.path(['some', 'nice', 'path']),
  R.map(R.identity)
)(xs)
// where xs = {some: {not: {soNice: 'path'}}}
// => []
```

I think it's better to make the possibility of failure explicit and blow up as early as possible if data doesn't match programmer's expectations, as David also put it.

```js
const R = require('ramda');
const S = require('sanctuary');

const orEmpty = S.fromMaybe([]);
const getStuff = R.pipe(
  S.gets(Array, ['some', 'nice', 'path']),
  orEmpty,
  R.map(R.toUpper)
);

const arr = [
  { some: { not: { soNice: 'path' } } },
  { some: { nice: { path: ['foo', 'bar', 'xyz'] } } }
];

R.map(getStuff, arr);
// [ [], [ 'FOO', 'BAR', 'XYZ' ] ]
```

However even without using a Maybe type, `safeMap` would be better than a self healing `map`.

**Comment by jon49** [comment]
I guess, since `map` can take so many different data types, my ideal would that it would just return `undefined` rather than throwing an exception. I think this would be a nice compromise. Then if you have a chain of events it could just keep passing `undefined` to the next function. E.g.,

```ts
R.pipe(
  R.path(['some', 'nice', 'path']), //=> undefined
  R.filter(...), //=> undefined
  R.map(...) //=> undefined
)({...})
```

Ideally we would know what all the data is before. But since JavaScript is a loosely typed language it makes it much more difficult. In other languages, like F#, you can have your `options` and it will yell at you and won't let you compile if you don't write it right.. But not in JavaScript (or rather, it is tedious in JavaScript). In F# it also namespaces the type, so if you are mapping an a list it would be `List.map` or `Sequence.map`, etc. So I guess that that would be another option &rarr; `R.Array.map` which could be safe returning an empty array if it isn't an array.

**Comment by jon49** [comment]
One more thing. Since I'm working with legacy systems I don't always have control over what data I receive. Many of the calls are SOAP calls which are terrible to try and work with. Or, if I'm trying to abstract something data and expect a certain input but the input is dirty.

**Comment by davidchambers** [comment]
> I'm working with legacy systems I don't always have control over what data I receive.

I'm in the same position. This was the primary motivation for the creation of [Sanctuary][1].

> ```javascript
> R.pipe(
>   R.path(['some', 'nice', 'path']), //=> undefined
>   R.filter(...), //=> undefined
>   R.map(...) //=> undefined
> )({...})
> ```

Instead, I would write:

```javascript
S.pipe([S.gets(Array, ['some', 'nice', 'path']),    // :: Maybe [*]
        R.map(R.filter(S.is(String))),              // :: Maybe [String]
        R.map(R.map(R.toUpper))],                   // :: Maybe [String]
       {...})
```


[1]: https://github.com/plaid/sanctuary

**Comment by jon49** [comment]
Yeah, that is a interesting way to do it. I've been using `TypeScript` and converting it to `JSON-Schema`. That way I get the static typing and ease of use, then convert it to `JSON-Schema` which gives me a bunch of tools, like Swagger, a `json-schema` to `mongoose` schema converter. I will eventually use json-schema (via Typescript and Swagger) to create my routing and testing -- that way a route isn't created without documentation and testing.

Now, my eventual goal will be to create a GUI that I can click and drag parts of the object around, rename different parts of the object, link to functions to manipulate different parts, validate, etc. I think it would be much easier to visually see what I am doing and compare the old to the new (especially when the object is large).

Here's my first naive version (note, I didn't update all the example but I did add a bit more functionality): https://gist.github.com/jon49/8dbe86251c425ee38f16
Which is similar to `R.evolve` but more powerful. But it will need to work on the json-schema which will then be used to change an object.
